tests/test_conduit_cli.py

```diff
@@ -0,0 +1,197 @@
+def test_get_cluster_arn(alias_session):
+    """Test that, given app and environment strings, get_cluster_arn returns the
+    arn of a cluster tagged with these strings."""
+
+    arn = boto3.client("ecs").create_cluster(
+        tags=[
+            {"key": "copilot-application", "value": "dbt-app"},
+            {"key": "copilot-environment", "value": "staging"},
+            {"key": "aws:cloudformation:logical-id", "value": "Cluster"},
+        ]
+    )["cluster"]["clusterArn"]
+
+    assert get_cluster_arn("dbt-app", "staging") == arn
+
+
+@patch("subprocess.call")
+def test_create_task(subprocess_call):
+    """Test that create_task runs the `copilot task run` command with expected
+    --app and --env flags."""
+
+    create_task("dbt-app", "staging", "arn::blah")
+
+    subprocess_call.assert_called_once_with(
+        "copilot task run -n dbtunnel --dockerfile Dockerfile --secrets DB_SECRET=arn::blah --app dbt-app --env staging",
+        shell=True,
+    )
+
+
+@mock_secretsmanager
+def test_get_postgres_secret_arn():
+    """Test that, given app and environment strings, get_postgres_secret_arn
+    returns the app's secret arn string."""
+
+    arn = boto3.client("secretsmanager").create_secret(
+        Name="/copilot/dbt-app/staging/secrets/POSTGRES", SecretString="secretivestring"
+    )["ARN"]
+
+    assert get_postgres_secret_arn("dbt-app", "staging") == arn
+
+
+@mock_ec2
+@mock_ecs
+def test_is_task_running():
+    """Test that, given a cluster arn string, is_task_running returns False
+    until an ecs task is running on an ec2 instance in that cluster."""
+
+    ecs_client = boto3.client("ecs")
+    arn = ecs_client.create_cluster(
+        tags=[
+            {"key": "copilot-application", "value": "dbt-app"},
+            {"key": "copilot-environment", "value": "staging"},
+            {"key": "aws:cloudformation:logical-id", "value": "Cluster"},
+        ]
+    )["cluster"]["clusterArn"]
+
+    assert not is_task_running(arn)
+
+    client = boto3.client("ec2")
+    images = client.describe_images(Owners=["amazon"])["Images"]
+    client.run_instances(ImageId=images[0]["ImageId"], MinCount=1, MaxCount=1)
+    instances = boto3.client("ec2").describe_instances()
+    instance_id = instances["Reservations"][0]["Instances"][0]["InstanceId"]
+    ec2 = boto3.resource("ec2")
+    instance = ec2.Instance(instance_id)
+    instance_id_document = json.dumps(ec2_utils.generate_instance_identity_document(instance))
+    ecs_client.register_container_instance(cluster=arn, instanceIdentityDocument=instance_id_document)
+    definition_arn = ecs_client.register_task_definition(
+        family="copilot-dbtunnel",
+        containerDefinitions=[
+            {"name": "test_container", "image": "test_image", "cpu": 100, "memory": 500, "essential": True}
+        ],
+    )["taskDefinition"]["taskDefinitionArn"]
+    ecs_client.run_task(cluster=arn, taskDefinition=definition_arn)
+
+    assert is_task_running(arn)
```

Comment:
@yusufsheiqh and I were looking at this together and as part of getting to grips with what it's doing, we kind of ended up doing a bit of a rewrite in our IDE.

in summary, the test was asserting on two, opposite, outcomes, so we split it into two tests, one for the expected `False` and one for the expected `True`.

Then we renamed some things to make it more explicit that we are working with mocked things and separated the tests into Arrange, Act and Assert blocks.

The result follows in massive suggestion!

```suggestion
@mock_ecs
def test_is_task_running_when_task_is_not_running():
    """Given an ECS Cluster ARN string, is_task_running should return False when the task is not running."""

    mocked_ecs_client = boto3.client("ecs")
    mocked_cluster_arn = mocked_ecs_client.create_cluster(
        tags=[
            {"key": "copilot-application", "value": "dbt-app"},
            {"key": "copilot-environment", "value": "staging"},
            {"key": "aws:cloudformation:logical-id", "value": "Cluster"},
        ]
    )["cluster"]["clusterArn"]

    assert not is_task_running(mocked_cluster_arn)


@mock_ec2
@mock_ecs
def test_is_task_running():
    """Given an ECS Cluster ARN string, is_task_running should return True when the task is running."""

    # Create mocked ECS Cluster
    mocked_ecs_client = boto3.client("ecs")
    mocked_cluster_arn = mocked_ecs_client.create_cluster(
        tags=[
            {"key": "copilot-application", "value": "dbt-app"},
            {"key": "copilot-environment", "value": "staging"},
            {"key": "aws:cloudformation:logical-id", "value": "Cluster"},
        ]
    )["cluster"]["clusterArn"]
    # Create mocked EC2 instance
    mocked_ec2_client = boto3.client("ec2")
    mocked_ec2_images = mocked_ec2_client.describe_images(Owners=["amazon"])["Images"]
    mocked_ec2_client.run_instances(ImageId=mocked_ec2_images[0]["ImageId"], MinCount=1, MaxCount=1)
    mocked_ec2_instances = boto3.client("ec2").describe_instances()
    mocked_ec2_instance_id = mocked_ec2_instances["Reservations"][0]["Instances"][0]["InstanceId"]
    mocked_ec2 = boto3.resource("ec2")
    mocked_ec2_instance = mocked_ec2.Instance(mocked_ec2_instance_id)
    mocked_instance_id_document = json.dumps(ec2_utils.generate_instance_identity_document(mocked_ec2_instance))
    # Attach mocked EC2 instance to the mocked ECS Cluster
    mocked_ecs_client.register_container_instance(cluster=mocked_cluster_arn, instanceIdentityDocument=mocked_instance_id_document)
    mocked_task_definition_arn = mocked_ecs_client.register_task_definition(
        family="copilot-dbtunnel",
        containerDefinitions=[
            {"name": "test_container", "image": "test_image", "cpu": 100, "memory": 500, "essential": True}
        ],
    )["taskDefinition"]["taskDefinitionArn"]

    mocked_ecs_client.run_task(cluster=mocked_cluster_arn, taskDefinition=mocked_task_definition_arn)

    assert is_task_running(mocked_cluster_arn)
```

Reply:
We were also wondering about extracting things like the recurring `mocked_ecs_client.create_cluster...` pieces out to helper methods to reduce duplication and perhaps make the test scenarios more readable.